File: sdks/nodejs/tests/unit/sdk.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { createDeck, DEFAULT_ROOT } from '../../src/index.js';

describe('@deckops/sdk', () => {
  let mock: MockAdapter;

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.restore();
  });

  it('uses the default root', () => {
    const deck = createDeck();
    expect(deck.root).toBe(DEFAULT_ROOT);
  });

  it('sends token and apiKey headers when creating a task', async () => {
    const deck = createDeck({
      root: 'http://localhost:3000/api',
      token: 'token-1',
      apiKey: 'key-1',
      spaceId: 'space-1',
    });

    mock.onPost('http://localhost:3000/api/tools/tasks').reply((config) => {
      expect(config.headers?.['X-Auth-Token']).toBe('token-1');
      expect(config.headers?.Authorization).toBe('Bearer key-1');
      expect(JSON.parse(String(config.data))).toMatchObject({
        spaceId: 'space-1',
        fileIds: ['file-1'],
        type: 'convertor.ppt2pdf',
        name: 'slides',
        params: {},
      });
      return [
        200,
        {
          id: 'task-1',
          spaceId: 'space-1',
          type: 'convertor.ppt2pdf',
          status: 'pending',
        },
      ];
    });

    const task = await deck.convertPptToPdf({ fileIds: ['file-1'], name: 'slides' });
    expect(task.id).toBe('task-1');
  });

  it('lists, gets, deletes, and waits for tasks', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });

    mock
      .onGet('http://localhost:3000/api/tools/tasks')
      .reply(200, [{ id: 'task-1', spaceId: 'space-1', type: 'image.ocr', status: 'pending' }], {
        'x-content-record-total': '1',
      });
    mock
      .onGet('http://localhost:3000/api/tools/tasks/task-1')
      .reply(200, { id: 'task-1', spaceId: 'space-1', type: 'image.ocr', status: 'completed' }, {
        'content-type': 'application/json',
      });
    mock.onDelete('http://localhost:3000/api/tools/tasks/task-1').reply(200);

    const list = await deck.tasks.list();
    expect(list.total).toBe(1);
    expect(list.tasks[0]?.id).toBe('task-1');

    const got = await deck.tasks.get('task-1');
    expect(got.status).toBe('completed');

    const waited = await deck.tasks.wait('task-1', { useEventStream: false, timeout: 5 });
    expect(waited.status).toBe('completed');

    await expect(deck.tasks.delete('task-1')).resolves.toBeUndefined();
  });

  it('downloads task results through ttask.down', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });

    mock.onGet('http://localhost:3000/api/tools/tasks/task-1/download').reply(200, [
      ['https://cdn.example.com/out.pdf', 1024, 'hash-1', { total: 12, w: 1920, h: 1080 }],
    ]);

    const result = await deck.ttask.down<'convertor.ppt2pdf'>('task-1');
    expect(result[0]?.[0]).toBe('https://cdn.example.com/out.pdf');
    expect(result[0]?.[3]?.total).toBe(12);
  });

  it('passes download type for generation-style task downloads', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });

    mock.onGet('http://localhost:3000/api/tools/tasks/task-gen/download').reply((config) => {
      expect(config.params).toEqual({ _type: 'pptx' });
      return [200, { downloadUrl: 'https://cdn.example.com/deck.pptx' }];
    });

    const result = await deck.tasks.down<'generation'>('task-gen', { type: 'pptx' });
    expect(result.downloadUrl).toContain('deck.pptx');
  });

  it('requests upload auth and returns deduplicated file ids', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });

    mock.onPost('http://localhost:3000/api/spaces/space-1/file/auth').reply((config) => {
      const body = JSON.parse(String(config.data));
      expect(body.name).toBe('a.txt');
      expect(body.bytes).toBe(3);
      expect(body.hash).toBe('900150983cd24fb0d6963f7d28e17f72');
      return [
        200,
        {
          id: 'file-1',
          key: 'files/a.txt',
          hash: body.hash,
          platform: 'oss',
          multipart: false,
        },
      ];
    });

    const result = await deck.files.upload(new Uint8Array([97, 98, 99]), {
      name: 'a.txt',
    });
    expect(result.id).toBe('file-1');
  });

  it('uploads files inside task helpers before creating tasks', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });

    mock.onPost('http://localhost:3000/api/spaces/space-1/file/auth').reply((config) => {
      const body = JSON.parse(String(config.data));
      expect(body.name).toBe('slides.pptx');
      expect(body.bytes).toBe(3);
      expect(body.hash).toBe('900150983cd24fb0d6963f7d28e17f72');
      return [
        200,
        {
          id: 'uploaded-file-1',
          key: 'files/slides.pptx',
          hash: body.hash,
          platform: 'oss',
          multipart: false,
        },
      ];
    });

    mock.onPost('http://localhost:3000/api/tools/tasks').reply((config) => {
      expect(JSON.parse(String(config.data))).toMatchObject({
        spaceId: 'space-1',
        fileIds: ['existing-file', 'uploaded-file-1'],
        type: 'convertor.ppt2pdf',
      });
      return [
        200,
        {
          id: 'task-1',
          spaceId: 'space-1',
          type: 'convertor.ppt2pdf',
          status: 'pending',
          fileIds: ['existing-file', 'uploaded-file-1'],
        },
      ];
    });

    const task = await deck.convertPptToPdf({
      fileIds: ['existing-file'],
      files: [{ input: new Uint8Array([97, 98, 99]), name: 'slides.pptx' }],
    });

    expect(task.fileIds).toEqual(['existing-file', 'uploaded-file-1']);
  });

  it('calculates hashes for Blob uploads', async () => {
    const deck = createDeck({ root: 'http://localhost:3000/api', token: 'token-1', spaceId: 'space-1' });
    const file = new File([new Uint8Array([97, 98, 99])], 'browser.txt');

    mock.onPost('http://localhost:3000/api/spaces/space-1/file/auth').reply((config) => {
      const body = JSON.parse(String(config.data));
      expect(body.name).toBe('browser.txt');
      expect(body.bytes).toBe(3);
      expect(body.hash).toBe('900150983cd24fb0d6963f7d28e17f72');
      return [
        200,
        {
          id: 'browser-file-1',
          key: 'files/browser.txt',
          hash: body.hash,
          platform: 'oss',
          multipart: false,
        },
      ];
    });

    const result = await deck.files.upload(file);
    expect(result.id).toBe('browser-file-1');
  });

  it('retries once after onUnauthorized updates credentials', async () => {
    const deck = createDeck({
      root: 'http://localhost:3000/api',
      token: 'old-token',
      spaceId: 'space-old',
      onUnauthorized: async () => ({ token: 'new-token', spaceId: 'space-new' }),
    });

    mock.onGet('http://localhost:3000/api/tools/tasks/task-1').replyOnce(401, { message: 'expired' });
    mock.onGet('http://localhost:3000/api/tools/tasks/task-1').reply((config) => {
      expect(config.headers?.['X-Auth-Token']).toBe('new-token');
      expect(config.params).toEqual({ spaceId: 'space-new' });
      return [
        200,
        { id: 'task-1', spaceId: 'space-new', type: 'image.ocr', status: 'completed' },
        { 'content-type': 'application/json' },
      ];
    });

    const task = await deck.tasks.get('task-1');
    expect(task.spaceId).toBe('space-new');
  });
});

File: sdks/go/sdk_test.go
package deckops

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
	"time"
)

const testAuthUUID = "f47ac10b-58cc-4372-a567-0e02b2c3d479"

type memoryUUIDStorage struct {
	value string
}

func (s *memoryUUIDStorage) Get(context.Context) (string, error) {
	return s.value, nil
}

func (s *memoryUUIDStorage) Set(_ context.Context, value string) error {
	s.value = value
	return nil
}

func TestDefaultRoot(t *testing.T) {
	resetAuthUUIDCacheForTests()
	deck, err := New(context.Background(), ClientOptions{AuthUUID: testAuthUUID})
	if err != nil {
		t.Fatal(err)
	}
	if deck.Root() != DefaultRoot {
		t.Fatalf("root = %q, want %q", deck.Root(), DefaultRoot)
	}
}

func TestCreateTaskSendsAuthHeaders(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/tools/tasks" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		if got := r.Header.Get("X-Auth-Token"); got != "token-1" {
			t.Fatalf("token header = %q", got)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer key-1" {
			t.Fatalf("authorization = %q", got)
		}
		if got := r.Header.Get("X-Auth-UUID"); got != testAuthUUID {
			t.Fatalf("auth uuid = %q", got)
		}
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body["spaceId"] != "space-1" || body["type"] != string(TaskConvertPptToPDF) || body["name"] != "slides" {
			t.Fatalf("unexpected body: %#v", body)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"id":"task-1","spaceId":"space-1","type":"convertor.ppt2pdf","status":"pending"}`))
	}))
	defer server.Close()

	deck, err := New(ctx, ClientOptions{
		Root:     server.URL,
		Token:    "token-1",
		APIKey:   "key-1",
		SpaceID:  "space-1",
		AuthUUID: testAuthUUID,
	})
	if err != nil {
		t.Fatal(err)
	}
	task, err := deck.ConvertPptToPDF(ctx, TaskShortcutParams{FileIDs: []string{"file-1"}, Name: "slides"})
	if err != nil {
		t.Fatal(err)
	}
	if task.ID != "task-1" {
		t.Fatalf("task id = %q", task.ID)
	}
}

func TestListGetDeleteAndWait(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch {
		case r.Method == http.MethodGet && r.URL.Path == "/tools/tasks":
			w.Header().Set("X-Content-Record-Total", "1")
			_, _ = w.Write([]byte(`[{"id":"task-1","spaceId":"space-1","type":"image.ocr","status":"pending"}]`))
		case r.Method == http.MethodGet && r.URL.Path == "/tools/tasks/task-1":
			_, _ = w.Write([]byte(`{"id":"task-1","spaceId":"space-1","type":"image.ocr","status":"completed"}`))
		case r.Method == http.MethodDelete && r.URL.Path == "/tools/tasks/task-1":
			w.WriteHeader(http.StatusOK)
		default:
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
	}))
	defer server.Close()

	deck, err := New(ctx, ClientOptions{Root: server.URL, SpaceID: "space-1", AuthUUID: testAuthUUID})
	if err != nil {
		t.Fatal(err)
	}
	list, err := deck.Tasks.List(ctx, ListTasksParams{})
	if err != nil {
		t.Fatal(err)
	}
	if list.Total != 1 || list.Tasks[0].ID != "task-1" {
		t.Fatalf("unexpected list: %#v", list)
	}
	got, err := deck.Tasks.Get(ctx, "task-1", false)
	if err != nil {
		t.Fatal(err)
	}
	if got.Status != TaskStatusCompleted {
		t.Fatalf("status = %s", got.Status)
	}
	waited, err := deck.Tasks.Wait(ctx, "task-1", WaitForTaskOptions{DisableSSE: true, Timeout: time.Second})
	if err != nil {
		t.Fatal(err)
	}
	if waited.Status != TaskStatusCompleted {
		t.Fatalf("waited status = %s", waited.Status)
	}
	if err := deck.Tasks.Delete(ctx, "task-1"); err != nil {
		t.Fatal(err)
	}
}

func TestUploadDeduplicatedFile(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/spaces/space-1/file/auth" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body["name"] != "a.txt" || int(body["bytes"].(float64)) != 3 || body["hash"] != "900150983cd24fb0d6963f7d28e17f72" {
			t.Fatalf("unexpected auth body: %#v", body)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"id":"file-1","key":"files/a.txt","hash":"900150983cd24fb0d6963f7d28e17f72","platform":"oss","multipart":false}`))
	}))
	defer server.Close()

	deck, err := New(ctx, ClientOptions{Root: server.URL, SpaceID: "space-1", AuthUUID: testAuthUUID})
	if err != nil {
		t.Fatal(err)
	}
	result, err := deck.Files.Upload(ctx, UploadInput{Name: "a.txt", Data: []byte("abc")}, UploadOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if result.ID != "file-1" {
		t.Fatalf("file id = %q", result.ID)
	}
}

func TestTaskHelperUploadsBeforeCreate(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/spaces/space-1/file/auth":
			_, _ = w.Write([]byte(`{"id":"uploaded-file-1","key":"files/slides.pptx","hash":"900150983cd24fb0d6963f7d28e17f72","platform":"oss","multipart":false}`))
		case "/tools/tasks":
			var body map[string]any
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				t.Fatal(err)
			}
			fileIDs := body["fileIds"].([]any)
			if len(fileIDs) != 2 || fileIDs[0] != "existing-file" || fileIDs[1] != "uploaded-file-1" {
				t.Fatalf("unexpected fileIds: %#v", fileIDs)
			}
			_, _ = w.Write([]byte(`{"id":"task-1","spaceId":"space-1","type":"convertor.ppt2pdf","status":"pending","fileIds":["existing-file","uploaded-file-1"]}`))
		default:
			t.Fatalf("unexpected path = %s", r.URL.Path)
		}
	}))
	defer server.Close()

	deck, err := New(ctx, ClientOptions{Root: server.URL, SpaceID: "space-1", AuthUUID: testAuthUUID})
	if err != nil {
		t.Fatal(err)
	}
	task, err := deck.ConvertPptToPDF(ctx, TaskShortcutParams{
		FileIDs: []string{"existing-file"},
		Files: []TaskUploadInput{{
			Input:   UploadInput{Data: []byte("abc")},
			Options: UploadOptions{Name: "slides.pptx"},
		}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(task.FileIDs) != 2 {
		t.Fatalf("file ids = %#v", task.FileIDs)
	}
}

func TestUnauthorizedRefreshesCredentials(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	var calls int
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		w.Header().Set("Content-Type", "application/json")
		if calls == 1 {
			w.WriteHeader(http.StatusUnauthorized)
			_, _ = w.Write([]byte(`{"message":"expired"}`))
			return
		}
		if got := r.Header.Get("X-Auth-Token"); got != "new-token" {
			t.Fatalf("token = %q", got)
		}
		if got := r.URL.Query().Get("spaceId"); got != "space-new" {
			t.Fatalf("spaceId = %q", got)
		}
		_, _ = w.Write([]byte(`{"id":"task-1","spaceId":"space-new","type":"image.ocr","status":"completed"}`))
	}))
	defer server.Close()

	deck, err := New(ctx, ClientOptions{
		Root:     server.URL,
		Token:    "old-token",
		SpaceID:  "space-old",
		AuthUUID: testAuthUUID,
		OnUnauthorized: func(context.Context) (AuthRefresh, error) {
			return AuthRefresh{Token: "new-token", SpaceID: "space-new"}, nil
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := deck.Tasks.Get(ctx, "task-1", false); err != nil {
		t.Fatal(err)
	}
}

func TestAuthUUIDStorage(t *testing.T) {
	resetAuthUUIDCacheForTests()
	ctx := context.Background()
	storage := &memoryUUIDStorage{}
	deck, err := New(ctx, ClientOptions{AuthUUIDStorage: storage})
	if err != nil {
		t.Fatal(err)
	}
	if !IsValidAuthUUID(deck.AuthUUID()) {
		t.Fatalf("invalid auth uuid: %q", deck.AuthUUID())
	}
	if storage.value != deck.AuthUUID() {
		t.Fatalf("stored = %q, uuid = %q", storage.value, deck.AuthUUID())
	}

	again, err := New(ctx, ClientOptions{AuthUUIDStorage: storage})
	if err != nil {
		t.Fatal(err)
	}
	if again.AuthUUID() != storage.value {
		t.Fatalf("uuid was not reused")
	}
}

func TestMain(m *testing.M) {
	oldConfigDir := os.Getenv("DECKOPS_CONFIG_DIR")
	oldAuthUUID := os.Getenv("DECKOPS_AUTH_UUID")
	tempDir, err := os.MkdirTemp("", "deckops-go-sdk-test-*")
	if err != nil {
		panic(err)
	}
	_ = os.Setenv("DECKOPS_CONFIG_DIR", tempDir)
	_ = os.Unsetenv("DECKOPS_AUTH_UUID")
	code := m.Run()
	if oldConfigDir == "" {
		_ = os.Unsetenv("DECKOPS_CONFIG_DIR")
	} else {
		_ = os.Setenv("DECKOPS_CONFIG_DIR", oldConfigDir)
	}
	if oldAuthUUID == "" {
		_ = os.Unsetenv("DECKOPS_AUTH_UUID")
	} else {
		_ = os.Setenv("DECKOPS_AUTH_UUID", oldAuthUUID)
	}
	_ = os.RemoveAll(tempDir)
	os.Exit(code)
}

func TestParseSSE(t *testing.T) {
	var got Task
	err := parseSSE(strings.NewReader("data: {\"id\":\"task-1\",\"status\":\"completed\"}\n\n"), func(data string) bool {
		if err := json.Unmarshal([]byte(data), &got); err != nil {
			t.Fatal(err)
		}
		return got.Status == TaskStatusCompleted
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "task-1" {
		t.Fatalf("task id = %q", got.ID)
	}
}
